### ISSUE TITLE
Add group support to the 'send' REST endpoint

### ISSUE THREAD

**Issue by wpeckr** [bbernhard/signal-cli-rest-api#12]
https://github.com/AsamK/signal-cli/blob/master/man/signal-cli.1.adoc#send

**Comment by rgruyters** [comment]
That would be a nice additional feature.

**Comment by bbernhard** [comment]
Is there anyone who is interested in helping to test this feature? I think most of the functionality is done now, but I don't want to merge it into the main docker image until some more testing. So, if anyone wants to help, please let me know. I can push a docker image for testing to dockerhub

**Comment by Loki321** [comment]
Happy to test.

**Comment by mchl-btlr** [comment]
It had already been working before, I have just prefixed "-g " before the group id. But native support is clearly better. Woud also be happy to test.

**Comment by bbernhard** [comment]
@Loki321 @mchl-btlr very cool, thanks!

the docker image can be pulled via: `docker pull bbernhard/signal-cli-rest-api:latest-dev`. If you are using the a docker-compose file similar to the one in the README.md file, just change the `image` parameter to the dev image (`bbernhard/signal-cli-rest-api:latest-dev`). 

The new APIs are described [here](https://github.com/bbernhard/signal-cli-rest-api/tree/groups) in the README.md. The API supports creating groups, listing groups, deleting groups and sending a message to a group.

**Comment by Loki321** [comment]
@bbernhard excellent. I have run through each API call and they are working as expected here.

There isn't an example given for how to add an attachment to a group message but it also works as it should.

I'm using this mainly for home-assistant notifications, I guess an update to the notification component is necessary to take advantage of sending to a group from home-assistant?

**Comment by bbernhard** [comment]
Thanks for testing!

> I'm new to signal in general so am unsure how it is supposed to behave when deleting a group but when deleting a group through this API it removes itself from the group but the group still exists with the other members still active. I expected the group to disappear for everyone but maybe this is the expected behaviour?

To be honest, this behavior came also a bit surprising to me. But I do not have any experiences with signal groups either - that was the first time I played with the groups feature. :smile: 

 > I'm using this mainly for home-assistant notifications, I guess an update to the notification component is necessary to take advantage of sending to a group from home-assistant?

yeah, right - unfortunately this also needs some changes in the Home Assistant codebase. If you would like to have that feature in Home Assistant, please create a ticket in the [Home Assistant repository](https://github.com/home-assistant/core), so that we have that properly tracked. As I am the code owner for the signal messenger integration, I should then automatically get the ticket assigned :)

**Comment by Loki321** [comment]
No problem, thank you for creating it!

I have now moved all of my notifications over (Unraid,Plex,etc..) and it's working great.

> yeah, right - unfortunately this also needs some changes in the Home Assistant codebase. If you would like to have that feature in Home Assistant, please create a ticket in the [Home Assistant repository](https://github.com/home-assistant/core), so that we have that properly tracked. As I am the code owner for the signal messenger integration, I should then automatically get the ticket assigned :)

Done.

**Comment by bbernhard** [comment]
> Done.

Just saw that your feature request got closed - sorry, wasn't aware that they don't use the ticket system for feature requests. I'll try to find time to implement that feature soon and will create a PR in the Home Assistant repo.

**Comment by rgruyters** [comment]
Been testing it and so far it works fine. No issues yet.

**Comment by bbernhard** [comment]
Thanks for testing! If there are no issues raised in the next days, I'll merge the changes in the stable docker image. 

Regarding the Home Assistant integration: 

At the moment the `configuration.yaml` for the signal messenger integration looks like this: 

```
notify:
  - name: signal
    platform: signal_messenger
    url: "http://127.0.0.1:8080"
    number: "<number>"
    recipients:
      - "<recipient1>
      - "<recipient2>
      - ....
```

Now, with the group messages we need to enrich the `configuration.yaml` with a `group_id` parameter somehow. Unfortunately, the `recipients` and the `group_id` are mutually exclusive - you can't specify both. 

For sending group messages I thought about something like this: 

```
notify:
  - name: signal
    platform: signal_messenger
    url: "http://127.0.0.1:8080"
    number: "<number>"
    group_id: "<group id>"
```

Does that make sense? Do you have other suggestions or would you name the parameter differently? I am myself not a Home Assistant user that uses a lot of different integrations - so I don't know if there are other integrations out there with mutually exclusive parameters. Would be interesting to see how those integrations handle it. 

If anyone of you @wpeckr @rgruyters @Loki321 @mchl-btlr is a Home Assistant user and has some suggestions, please let me know.

**Comment by Loki321** [comment]
Makes sense, or maybe;
```
  - name: signal
    platform: signal_messenger
    url: "http://127.0.0.1:8080"
    group: yes|no (defaults to no)
    number: "<number>"
    recipients:
      - "<group_id>"

**Comment by bbernhard** [comment]
Many thanks! I really like that - the only small change I did was to change it to `is_group: true|false`. 

I've also changed the REST API endpoint for sending group messages a bit - it now also used the `is_group` parameter instead of the explicit `group_id` parameter. The other endpoints haven't changed. I've also updated the [API documentation](https://github.com/bbernhard/signal-cli-rest-api/tree/groups), in case someone wants to have a look. At the moment, it's only allowed to use one `group id` in the `recipients` list - the restriction is there, because the underlying library (`signal-cli` also only allows one `group id`). But in case there's demand for sending messages to multiple groups, it can easily be fix that in the backend (just need to invoke the `signal-cli` binary multiple times) without breaking the API. 

btw: in case someone wants to test the changes, I've also updated the `bbernhard/signal-cli-rest-api:latest-dev` docker image. 

If there are no objections, I'll merge the changes into the `stable` image soon and create a pull request for the groups feature in the Home Assistant repo. Hope to get that into the next 0.109 release :crossed_fingers:
